core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -366,8 +379,27 @@ class BrokerLifecycleManager(
       new BrokerRegistrationResponseHandler())
   }
 
+  // the response handler is not invoked from the event handler thread,
+  // so it is not safe to update state here, instead, schedule an event
+  // to continue handling the response on the event handler thread
   private class BrokerRegistrationResponseHandler extends ControllerRequestCompletionHandler {
     override def onComplete(response: ClientResponse): Unit = {
+      eventQueue.prepend(new BrokerRegistrationResponseEvent(response, false))
+    }
+
+    override def onTimeout(): Unit = {
+      info(s"Unable to register the broker because the RPC got timed out before it could be sent.")
+      eventQueue.prepend(new BrokerRegistrationResponseEvent(null, true))
+    }
+  }
+
+  private class BrokerRegistrationResponseEvent(response: ClientResponse, timedOut: Boolean) extends EventQueue.Event {
+    override def run(): Unit = {
+      communicationInFlight = false
```

Comment:
It seems that we don't need this since `communicationInFlight` is only set to true when sending a `HeartbeatRrequest`?

Reply:
I don't think so. We set `communicationInFlight = true` in `CommunicationEvent.run`, which sends both heartbeats but also registration requests.

Maybe I'm missing something?

Reply:
Since `NodeToControllerChannelManagerImpl` uses 1 for `maxInFlightRequestsPerConnection`, we could also add a `sendRequest` api in `NodeToControllerChannelManager`, instead 

networkClient.ready(node, now)

Reply:
Ok. I wasn't sure that communicationInFlight is for both requests.

Since `NodeToControllerChannelManagerImpl` uses 1 for `maxInFlightRequestsPerConnection`, another option is to add a `canSendRequest` api in `NodeToControllerChannelManager`, instead maintaining `communicationInFlight` here. 

`canSendRequest` api can be implemented through `networkClient.ready(node, now)`.